docs/design/2290-cadence-ndc.md

```diff
@@ -0,0 +1,310 @@
+# Design doc: Cadence NDC
+
+Last updated: Dec 2019
+
+Reference: [#2290](https://github.com/uber/cadence/issues/2290)
+
+
+## Abstract
+
+Cadence Cross DC is a feature which asynchronously replicates workflows from active data center to other passive data centers, for reconstruction. When necessary, customer can failover to any of the data centers which has the backup for high availability.
+
+Cadence Cross DC is an AP (in terms of CAP).
+
+This doc explains the high level concepts about Cadence Cross DC (mainly the N data center version).
+
+
+## Newly Introduced Components
+
+1. Version
+2. Version history
+3. Workflow history conflict resolution
+4. Zombie workflows
+5. Workflow task processing
+
+### Version
+
+Version is a newly introduced concept in Cadence Cross DC which describes the chronological order of events (per customer domain).
+
+Cadence Cross DC is AP, all domain change events & workflow history events are replicated asynchronously for high throughput. This means that data across data centers are not strongly consistent. To guarantee that domain data & workflow data will achieve eventual consistency (expecially when there is data conflict during a failover), version is introduced and attached to customers' domains. All workflow history events generated in a domain will also come with the version in that domain.
+
+All participating data centers are pre-configured with a unique initial version, and a shared version increment:
+
+  `initial version < shared version increment`
+
+When performing failover for one domain from one data center to another data center, the version attched to the domain will be changed by the following rule:
+
+  for all versions which follow `version % (shared version increment) == (active data centers' initial version)`
+  find the smallest version which has `version >= old version in domain`
+
+When there is a data conflict, comparison will be made and workflow history events with the highest version will win.
+
+When a data center is trying to mutate a workflow, version will be checked. A data center can mutate a workflow if and only if
+
+  1. version in the domain belongs to this data center, i.e.
+  `(vesion in domain) % (shared version increment) == (this data centers' initial version)`
+  2. the version of this workflow's last event is not less then version in domain, i.e.
+  `(last event's version) <= (version in domain)`
```

Comment:
```>=```

Reply:
if 1 is satisfied, what is the case that can lead to last event's version > version in domain?

Reply:
all replication tasks are handled asynchronously, meaning that domain replication (change of version) can experience delay.